crates/subspace-service/src/sync_from_dsn.rs

```diff
@@ -251,19 +290,70 @@ where
         .saturating_sub(chain_constants.confirmation_depth_k().into());
     let segment_header_downloader = SegmentHeaderDownloader::new(node);
 
-    while let Some(reason) = notifications.next().await {
-        pause_sync.store(true, Ordering::Release);
+    let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+
+    pause_sync.store(true, Ordering::Release);
+
+    let finalized_hash_existed = client.info().finalized_hash != client.info().genesis_hash;
+
+    if fast_sync_enabled && finalized_hash_existed {
+        debug!("Fast sync detected existing finalized hash.");
+    }
+
+    // Run fast-sync first.
+    #[allow(clippy::collapsible_if)] // clippy error: the if statement is not the same
+    if fast_sync_enabled && !finalized_hash_existed {
+        if notifications.next().await.is_some() {
+            let fast_syncer = FastSyncer::new(
+                &segment_headers_store,
+                node,
+                piece_getter,
+                client.clone(),
+                import_queue_service.clone(),
+                network_service.clone(),
+                fast_sync_state,
+                sync_service,
+            );
 
+            let fast_sync_result = fast_syncer.sync().await;
+
+            match fast_sync_result {
+                Ok(fast_sync_result) => {
+                    if fast_sync_result.skipped {
+                        debug!("Fast sync was skipped.");
+                    } else {
+                        last_processed_block_number = fast_sync_result.last_imported_block_number;
+                        last_processed_segment_index = fast_sync_result.last_imported_segment_index;
+                        reconstructor = fast_sync_result.reconstructor;
+
+                        debug!(%last_processed_block_number, %last_processed_segment_index, "Fast sync finished.");
+                    }
+                }
+                Err(err) => {
+                    error!("Fast sync failed: {err}");
+                }
+            }
+
+            notifications_sender
+                .try_send(NotificationReason::WentOnlineSubspace)
+                .map_err(|_| {
+                    sc_service::Error::Other("Can't send sync notification reason.".into())
+                })?;
+        }
+    }
```

Comment:
As discussed before this doesn't seem to belong to full sync from DSN, right now it is a one-off separate sync that happens before DSN sync . DSN sync is not a continuation of this sync because we will have new enough state for regular sync to kick in and take it from there, no need to pass reconstructor through everything.

Reply:
> I added the shared reconstructor when I encountered an error

You know the next natural question I'll have would be "which error did you encounter exactly"?

> Why do you think there is no need to pass the reconstructor? I appreciate more details here.

Everything worked fine before, so it is not on me to prove that with new code added this is not necessary, you should convince me that something that didn't exist before is now necessary (which is very possible, I just don't see it right now).

> There are two more reasons to place it here: fast-sync IS DSN-sync, shared sync pause management. However, I'm open to suggestions here.

Arguably it is not. Fast sync's primary goal is to import a single block from DSN: the very first block of the last segment. Even if node shuts down right after and DSN sync fails after restart, we can still continue regular Substrate sync successfully. The fact that we import other blocks from the last segment is an optimization (because we already have it, no need to throw them away), not a requirement.

And I already mentioned in one of the prior discussions that if we download source pieces of last segment sequentially we can both continue downloading last segment and import blocks concurrently before we've got the whole segment. It is important to distinguish what is the goal and the core feature we are trying to build and what are "nice to have" on top.

Reply:
> You know the next natural question I'll have would be "which error did you encounter exactly"?

It's hard to reproduce it on the fly but AFAIR the new reconstructor will result in a missing block on the edges of two segments (the last imported from the fast sync and the remaining segment obtained through DSN sync).

Reply:
DSN sync is designed to resume from any point, it'll just have to potentially download some of the data twice, but as articulated a few times already we don't need to call DSN sync after fast sync in the first place because regular Substrate sync can pick up from there already.